consensus/src/common.rs

```diff
@@ -103,88 +77,93 @@ impl Chain<ProtocolStateLegacy> for ProtocolStateChain {
-        self.0.first()
+    /// he function returns the consensus state of a block or chain.
+    /// The input is a block or chain X and the output is the consensus state.
+    pub fn consensus_state(&self) -> Option<&ConsensusState> {
+        self.top().map(|s| s.consensus_state())
     }
 
-    fn consensus_state(&self) -> Option<&ConsensusState> {
-        self.top().map(|s| &s.body.consensus_state)
+    /// Gets ProtocolState of the genesis block
+    pub fn genesis_block(&self) -> Option<&T> {
+        self.0.first()
     }
 
-    fn global_slot(&self) -> Option<&GlobalSlot> {
-        self.top().map(|s| &s.body.consensus_state.curr_global_slot)
+    /// The function returns the global slot number of a chain or block.
+    /// The input X is either a chain or block and the output is the global slot number.
+    pub fn global_slot(&self) -> Option<&GlobalSlot> {
+        self.top().map(|s| &s.consensus_state().curr_global_slot)
     }
 
-    fn epoch_slot(&self) -> Option<u32> {
+    /// The function computes the epoch slot number of a block.
+    /// The output is the epoch slot number in [0, slots_per_epoch].
+    pub fn epoch_slot(&self) -> Option<u32> {
         self.global_slot()
             .map(|s| (s.slot_number.0 % s.slots_per_epoch.0))
     }
 
-    fn length(&self) -> usize {
+    /// The function the length of a chain. The output is the length of the chain in blocks.
+    pub fn length(&self) -> usize {
         self.consensus_state()
             .map(|s| s.blockchain_length.0 as usize)
             .unwrap_or(0)
     }
 
-    fn last_vrf_hash_digest(&self) -> Result<String, ConsensusError> {
+    /// This function returns the hex digest of the hash of the last VRF output
+    ///  of a given chain. The input is a chain C and the output is the hash digest.
+    pub fn last_vrf_hash_digest(&self) -> Result<String, ConsensusError> {
         let hash = self
-            .top()
+            .consensus_state()
             .ok_or(ConsensusError::TopBlockNotFound)?
-            .body
-            .consensus_state
             .last_vrf_output
             .digest();
         Ok(hex::encode(hash))
     }
 
-    fn state_hash(&self) -> Option<Fp> {
+    /// This function returns hash of the top block's protocol state for a given chain.
+    /// The input is a chain C and the output is the hash.
+    pub fn state_hash(&self) -> Option<Fp> {
         self.top().map(|s| s.state_hash_fp())
     }
 }
 
-/// A trait that defines operations for consensus calculation
-pub trait Consensus {
-    /// Chain type
-    type Chain;
+/// A trait that defines operations for chain selection
+pub trait ChainSelection {
     /// Top level API to select between chains during a fork.
-    fn select_secure_chain<'a>(
-        &'a self,
-        candidates: &'a [Self::Chain],
-    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+    fn select_secure_chain(&mut self, candidates: Vec<Self>) -> Result<(), ConsensusError>
```

Comment:
Using borrows is sufficient for writing tests but not for real-world cases. IMO it's better to take over the ownership to avoid unnecessary clones

Reply:
And it makes it much easier for users if this API takes mutable self and makes in-place updates, as is shown in `naive_transition_frontier.rs`